lib/phoenix_live_view/async.ex

```diff
@@ -3,13 +3,83 @@ defmodule Phoenix.LiveView.Async do
 
   alias Phoenix.LiveView.{AsyncResult, Socket, Channel}
 
+  if Application.compile_env(:phoenix_live_view, :warn_on_expensive_assigns, false) do
+    defp validate_function_env(func, :start_async) do
+      {:env, variables} = Function.info(func, :env)
+
+      if Enum.any?(variables, &match?(%Phoenix.LiveView.Socket{}, &1)) do
+        IO.warn("""
+        you are accessing the LiveView Socket inside a start_async function.
+
+        This is an expensive operation because the whole socket is copied to the new process.
+
+        Instead of:
+
+            start_async(socket, :key, fn ->
+              do_something(socket.assigns.my_assign)
+            end)
+
+        You should do:
+
+            my_assign = socket.assigns.my_assign
+
+            start_async(socket, :key, fn ->
+              do_something(my_assign)
+            end)
+
+        For more information, see https://hexdocs.pm/elixir/1.16.1/process-anti-patterns.html#sending-unnecessary-data.
+        """)
+      end
+    end
+
+    defp validate_function_env(func, :assign_async) do
+      {:env, variables} = Function.info(func, :env)
+
+      if Enum.any?(variables, &match?(%Phoenix.LiveView.Socket{}, &1)) do
+        IO.warn("""
+        you are accessing the LiveView Socket inside an assign_async function.
+
+        This is an expensive operation because the whole socket is copied to the new process.
+
+        Instead of:
+
+            assign_async(socket, :key, fn ->
+              result = do_something(socket.assigns.my_assign)
+              {:ok, %{key: result}}
+            end)
+
+        You should do:
+
+            my_assign = socket.assigns.my_assign
+
+            assign_async(socket, :key, fn ->
+              result = do_something(my_assign)
+              {:ok, %{key: result}}
+            end)
+
+        For more information, see https://hexdocs.pm/elixir/1.16.1/process-anti-patterns.html#sending-unnecessary-data.
+        """)
+      end
+    end
```

Comment:
```suggestion
    defp validate_function_env(func, op) do
      {:env, variables} = Function.info(func, :env)

      if Enum.any?(variables, &match?(%Phoenix.LiveView.Socket{}, &1)) do
        IO.warn("""
        you are accessing the LiveView Socket inside a function given to #{op}.

        This is an expensive operation because the whole socket is copied to the new process.

        Instead of:

            #{op}(socket, :key, fn ->
              do_something(socket.assigns.my_assign)
            end)

        You should do:

            my_assign = socket.assigns.my_assign

           #{op}(socket, :key, fn ->
              do_something(my_assign)
            end)

        For more information, see https://hexdocs.pm/elixir/1.16.1/process-anti-patterns.html#sending-unnecessary-data.
        """)
      end
    end
```

Reply:
@josevalim that is not correct because the result of an assign_async is of the form `{:ok, %{my_assign: value}}`

Reply:
Hmm well I guess you could argue that the `do_something` function could return the correct format, so maybe it's fine?

Reply:
Yeah :)